src/components/cylc/graph/Graph.vue

```diff
@@ -0,0 +1,631 @@
+       * Graph data. These values are empty by default, then are populated later via watchers.
+       */
+      graphData: {
+        nodes: [],
+        edges: []
+      },
+      cytoscapeInstance: null
+    }
+  },
+
+  /**
+   * The watchers created by this component.
+   *
+   * The workflows property is mapped by Vuex (mapState). Whenever Vuex changes it, due to
+   * changes to the store, we are automatically notified. Even if the workflows object reference
+   * was not changed, but one of its properties was (deep: true).
+   *
+   * When the workflowUpdated method is executed, it will calculate the graphData, and change it
+   * once done.
+   *
+   * The graphData watcher will react to changes (not deep!) of graphData, and call the function
+   * updateGraph (debounced). This function is responsible for then - as per its name - update
+   * the graph, the layout, any other setting necessary once the nodes and edges have been added
+   * or removed.
+   */
+  watch: {
+    workflows: {
+      handler: function (newValue) {
+        this.workflowUpdated(newValue)
+      },
+      deep: true
+    },
+    graphData: function () {
+      this.updateGraph()
+    }
+  },
+
+  /**
+   * When this component is created, we simply create debounced functions
+   */
+  created () {
+    this.updateGraph = debounce(this.updateGraph_, 100)
+    this.workflowUpdated = debounce(this.workflowUpdated_, 100)
+  },
+
+  mounted () {
+    // create a div for cytoscape, this is an approach similar to vue-cytoscape
+    // but the vue-cytoscape component offers more than what we need, so we imitate
+    // part of its behaviour here, and ignore the rest and handle the graph and cytoscape
+    // settings ourselves.
+    const element = document.createElement('div')
+    element.setAttribute('class', 'cytoscape')
+    this.$el.appendChild(element)
+    this.cytoscapeInstance = Object.freeze(cytoscape(({
+      container: element
+    })))
+    this.registerExtensions()
+    cytoscape.use(dagre)
+    this.setupEventListeners()
+    this.runLayout(this.cytoscapeInstance)
+    this.setupUndoRedo(this.cytoscapeInstance)
+    this.setupPanzoom(this.cytoscapeInstance)
+    this.setupNavigator(this.cytoscapeInstance)
+    this.setupInteractivity(this.cytoscapeInstance)
+    this.setupHtmlLabel(this.cytoscapeInstance, states)
+    // set the initial zoom-level and pan (position) of the graph
+    // to set only the zoom level, call: this.cytoscapeInstance.zoom(0.5)
+    this.cytoscapeInstance.viewport({
+      zoom: 0.5,
+      pan: {
+        x: 100,
+        y: 100
+      }
+    })
+    this.loading = false
+  },
+
+  beforeDestroy () {
+    // before the component is destroyed, we need to remember to remove the tooltip, otherwise it is still visible
+    // in the next view/route.
+    if (this.tippy !== null) {
+      this.tippy.hide()
+    }
+    this.cytoscapeInstance.panzoom('destroy')
+    document.removeEventListener('keydown', this.keydownEventHandler)
```

Comment:
These were not being removed. So we had possibly some detached objects & memory leaking somewhere. (Found about the Panzoom clean up call while looking at their docs to set up default values, by accident).